lightning/src/offers/flow.rs

```diff
@@ -0,0 +1,1108 @@
+	pub fn verify_invoice_request(
+		&self, invoice_request: InvoiceRequest, context: Option<OffersContext>,
+	) -> Result<VerifiedInvoiceRequest, ()> {
+		let secp_ctx = &self.secp_ctx;
+		let expanded_key = &self.inbound_payment_key;
+
+		let nonce = match context {
+			None if invoice_request.metadata().is_some() => None,
+			Some(OffersContext::InvoiceRequest { nonce }) => Some(nonce),
+			_ => return Err(()),
+		};
+
+		let invoice_request = match nonce {
+			Some(nonce) => {
+				match invoice_request.verify_using_recipient_data(nonce, expanded_key, secp_ctx) {
+					Ok(invoice_request) => invoice_request,
+					Err(()) => return Err(()),
+				}
+			},
+			None => match invoice_request.verify_using_metadata(expanded_key, secp_ctx) {
+				Ok(invoice_request) => invoice_request,
+				Err(()) => return Err(()),
+			},
+		};
+
+		Ok(invoice_request)
+	}
+
+	/// Verifies a [`Bolt12Invoice`] using the provided [`OffersContext`] or the invoice's own metadata,
+	/// returning the corresponding [`PaymentId`] if successful.
+	///
+	/// - If an [`OffersContext::OutboundPayment`] with a `nonce` is provided, verification is performed
+	///   using the payer's context data.
+	/// - If no context is provided and the invoice corresponds to a [`Refund`] without blinded paths,
+	///   verification is performed using the invoice's metadata.
+	/// - If neither condition is met, verification fails.
+	pub fn verify_bolt12_invoice(
+		&self, invoice: &Bolt12Invoice, context: Option<&OffersContext>,
+	) -> Result<PaymentId, ()> {
+		let secp_ctx = &self.secp_ctx;
+		let expanded_key = &self.inbound_payment_key;
+
+		match context {
+			None if invoice.is_for_refund_without_paths() => {
+				invoice.verify_using_metadata(expanded_key, secp_ctx)
+			},
+			Some(&OffersContext::OutboundPayment { payment_id, nonce, .. }) => {
+				invoice.verify_using_payer_data(payment_id, nonce, expanded_key, secp_ctx)
+			},
+			_ => Err(()),
+		}
+	}
+
+	/// Verifies the provided [`AsyncPaymentsContext`] for an [`AsyncPaymentsMessage`].
+	///
+	/// Depending on whether the context is for an inbound or outbound payment, the function
+	/// performs verification using nonces and HMAC values, stored within the context.
+	///
+	/// - For **Inbound Payments**, the context is verified using the `nonce` and `hmac` values,
+	///   and ensures that the context has not expired based on `path_absolute_expiry`.
+	/// - For **Outbound Payments**, the context is verified using the `nonce` and `hmac` values,
+	///   and if valid, returns the associated [`PaymentId`].
+	///
+	/// # Errors
+	///
+	/// Returns `Err(())` if:
+	/// - The HMAC verification fails for either inbound or outbound context.
+	/// - The inbound payment context has expired.
+	///
+	/// [`AsyncPaymentsMessage`]: crate::onion_message::async_payments::AsyncPaymentsMessage
+	#[cfg(async_payments)]
+	pub fn verify_async_payments_context(
+		&self, context: AsyncPaymentsContext,
+	) -> Result<VerifiedAsyncPaymentContext, ()> {
```

Comment:
Can we split this method to having one method per async payments onion message and drop the `VerifiedAsyncPaymentContext` enum? It's a bit confusing that callers need to check that the `Ok` result is the right context.

Reply:
That makes sense! Removed the enum and separated the functions in [pr3639.14](https://github.com/shaavan/rust-lightning/commits/pr3639.14).